jetty-webapp/src/test/java/org/eclipse/jetty/webapp/ServletContextTmpAttributeTest.java

```diff
@@ -0,0 +1,176 @@
+    @Test
+    public void attributeWithNullValue() throws Exception
+    {
+        WebInfConfiguration webInfConfiguration = new WebInfConfiguration();
+        WebAppContext webAppContext = new WebAppContext();
+        webAppContext.setAttribute(ServletContext.TEMPDIR, null);
+        webInfConfiguration.resolveTempDirectory(webAppContext);
+        File tmp = webAppContext.getTempDirectory();
+        assertThat("webAppContext temp directory parent is java.io.tmpdir",
+                    tmp.getParentFile(),
+                    is(new File(System.getProperty("java.io.tmpdir"))));
+    }
+
+    /**
+     * ServletContext.TEMPDIR as String to valid directory
+     */
+    @Test
+    public void attributeWithStringValue() throws Exception
+    {
+        WebInfConfiguration webInfConfiguration = new WebInfConfiguration();
+        WebAppContext webAppContext = new WebAppContext();
+        Path tmp = Files.createTempDirectory("jetty_test");
+        webAppContext.setAttribute(ServletContext.TEMPDIR, tmp.toString());
+        webInfConfiguration.resolveTempDirectory(webAppContext);
+        File temp = webAppContext.getTempDirectory();
+        assertThat("webAppContext is the temp directory created",
+                   temp.toPath(),
+                   is(tmp));
+    }
+
+    /**
+     * ServletContext.TEMPDIR has <code>""</code> value
+     * IllegalStateException
+     */
+    @Test
+    public void attributeWithEmptyStringValue() throws Exception
+    {
+        WebInfConfiguration webInfConfiguration = new WebInfConfiguration();
+        WebAppContext webAppContext = new WebAppContext();
+        webAppContext.setAttribute(ServletContext.TEMPDIR, "");
+        assertThrows(IllegalStateException.class, () -> webInfConfiguration.resolveTempDirectory(webAppContext));
+    }
+
+    /**
+     * ServletContext.TEMPDIR as File to valid directory
+     */
+    @Test
+    public void attributeWithValidFileDirectoryValue() throws Exception
+    {
+        WebInfConfiguration webInfConfiguration = new WebInfConfiguration();
+        WebAppContext webAppContext = new WebAppContext();
+        Path tmp = Files.createTempDirectory("jetty_test");
+        webAppContext.setAttribute(ServletContext.TEMPDIR, tmp.toFile());
+        webInfConfiguration.resolveTempDirectory(webAppContext);
+        File temp = webAppContext.getTempDirectory();
+        assertThat("webAppContext is the temp directory created",
+                   temp.toPath(),
+                   is(tmp));
+    }
+
+    /**
+     * ServletContext.TEMPDIR as Path to valid directory
+     */
+    @Test
+    public void attributeWithValidPathDirectoryValue() throws Exception
+    {
+        WebInfConfiguration webInfConfiguration = new WebInfConfiguration();
+        WebAppContext webAppContext = new WebAppContext();
+        Path tmp = Files.createTempDirectory("jetty_test");
+        webAppContext.setAttribute(ServletContext.TEMPDIR, tmp);
+        // FIXME we should have an exception here
```

Comment:
We'd have to change WebInfConfiguration - AFAIK we've never supported a Path as the value of ServletContext.TEMPDIR.

Reply:
specs says it's a `java.io.File` but we support `String` which is wrong according to the specs....
changing WebInfConfiguration to have an exception or to support Path?

Reply:
Well, as we're already outside the spec, I guess we could accept Path too.

Reply:
I have implemented this.